metricbeat/module/windows/services/services_windows.go

```diff
@@ -0,0 +1,370 @@
+					displayNameOffset := uintptr(unsafe.Pointer(serviceTemp.LpDisplayName)) - (uintptr)(unsafe.Pointer(&servicesBuffer[0]))
+
+					displayNameBuffer.Reset()
+					serviceNameBuffer.Reset()
+
+					if err = sys.UTF16ToUTF8Bytes(servicesBuffer[displayNameOffset:serviceNameOffset], displayNameBuffer); err != nil {
+						return nil, err
+					}
+
+					if err = sys.UTF16ToUTF8Bytes(servicesBuffer[serviceNameOffset:lastOffset], serviceNameBuffer); err != nil {
+						return nil, err
+					}
+
+					lastOffset = displayNameOffset
+
+					service := ServiceStatus{}
+
+					service.DisplayName = displayNameBuffer.String()
+					service.ServiceName = serviceNameBuffer.String()
+
+					var state string
+
+					if stat, ok := serviceStates[ServiceState(serviceTemp.ServiceStatusProcess.DwCurrentState)]; ok {
+						state = stat
+					} else {
+						state = "Can not define State"
+					}
+					service.CurrentState = state
+
+					// Get detailed information
+					var serviceBufSize uint32
+					var serviceBytesNeeded uint32
+
+					serviceHandle, err := OpenService(handle, service.ServiceName, ServiceQueryConfig)
+					if err != nil {
+						return nil, err
+					}
+
+					if err := _QueryServiceConfig(serviceHandle, nil, serviceBufSize, &serviceBytesNeeded); err != nil {
+						if ServiceErrno(err.(syscall.Errno)) != SERVICE_ERROR_INSUFFICIENT_BUFFER {
+							if err := CloseServiceHandle(serviceHandle); err != nil {
+								return nil, err
+							}
+							return nil, err
+						}
+						serviceBufSize += serviceBytesNeeded
+						buffer := make([]byte, serviceBufSize)
+
+						for {
+							if err := _QueryServiceConfig(serviceHandle, &buffer[0], serviceBufSize, &serviceBytesNeeded); err != nil {
+								if ServiceErrno(err.(syscall.Errno)) != SERVICE_ERROR_INSUFFICIENT_BUFFER {
+									if err := CloseServiceHandle(serviceHandle); err != nil {
+										return nil, err
+									}
+									return nil, err
+								}
+								serviceBufSize += serviceBytesNeeded
+							} else {
+								serviceQueryConfig := (*QueryServiceConfig)(unsafe.Pointer(&buffer[0]))
+								service.StartType = serviceStartTypes[ServiceStartType(serviceQueryConfig.DwStartType)]
+								if err := CloseServiceHandle(serviceHandle); err != nil {
+									return nil, err
+								}
+								break
+							}
+						}
+					}
+
+					//Get uptime for service
+					if ServiceState(serviceTemp.ServiceStatusProcess.DwCurrentState) != ServiceStopped {
+
+						var processCreationTime syscall.Filetime
+						var processExitTime syscall.Filetime
+						var processKernelTime syscall.Filetime
+						var processUserTime syscall.Filetime
+
+						// Enable SeDebugPrivilege to opne processes
+
+						var token syscall.Token
+
+						currentProcess, err := syscall.GetCurrentProcess()
+						if err != nil {
+							return nil, err
+						}
+
+						if err := syscall.OpenProcessToken(currentProcess, syscall.TOKEN_ADJUST_PRIVILEGES, &token); err != nil {
```

Comment:
This probably should hard fail on a permission issue. I think it would be better to gracefully degrade such that the uptime info isn't available.